File: user-service/src/main/java/com/uniride/userservice/service/AmizadeService.java
package com.uniride.userservice.service;

import com.uniride.userservice.dto.response.AmizadeResponse;
import com.uniride.userservice.dto.response.PerfilPublicoResponse;
import com.uniride.userservice.dto.response.SaoAmigosResponse;

import java.util.List;

public interface AmizadeService {

    // Enviar solicitação de amizade
    AmizadeResponse enviarSolicitacao(Long solicitanteId, Long destinatarioId);

    // Responder solicitação
    AmizadeResponse aceitar(Long amizadeId, Long destinatarioId);
    AmizadeResponse recusar(Long amizadeId, Long destinatarioId);
    AmizadeResponse bloquear(Long amizadeId, Long destinatarioId);

    // Desfazer amizade aceita
    void remover(Long amizadeId, Long participanteId);

    // Cancelar solicitação pendente enviada pelo próprio usuário
    void cancelarSolicitacao(Long amizadeId, Long solicitanteId);

    // Listagens
    List<AmizadeResponse> listarAmigos(Long participanteId);
    List<AmizadeResponse> listarPendentesRecebidas(Long participanteId);
    List<AmizadeResponse> listarPendentesEnviadas(Long participanteId);
    List<AmizadeResponse> listarTodasEnviadas(Long participanteId);

    // Sugestões de amigos da mesma instituição
    List<PerfilPublicoResponse> sugestoesDaInstituicao(Long participanteId);

    // Busca por nome dentro da própria instituição (com statusAmizade embutido)
    List<PerfilPublicoResponse> buscarNaInstituicao(Long meuId, String termo);

    // Perfil público de qualquer participante (com statusAmizade embutido)
    PerfilPublicoResponse buscarPerfilPublico(Long participanteId, Long meuId);
    PerfilPublicoResponse buscarPerfilPublicoPorUsername(String username, Long meuId);

    // Lista amigos de outro usuário — retorna 403 se o requester não for amigo do target
    List<PerfilPublicoResponse> listarAmigosDeOutroUsuario(Long targetId, Long meuId);

    // Endpoint para o Feign Client do ride-service
    SaoAmigosResponse verificarAmizade(Long idA, Long idB);
}

File: mobile-backup/services/amizade.service.ts
import api from './api';
import { AmizadeResponse, ApiResponse, PerfilPublicoResponse } from '@/types/api.types';

export const amizadeService = {

  async listarAmigos(): Promise<AmizadeResponse[]> {
    const response = await api.get<ApiResponse<AmizadeResponse[]>>('/api/v1/amizades');
    return response.data.dados;
  },

  async listarPendentes(): Promise<AmizadeResponse[]> {
    const response = await api.get<ApiResponse<AmizadeResponse[]>>('/api/v1/amizades/pendentes');
    return response.data.dados;
  },

  async listarEnviadas(): Promise<AmizadeResponse[]> {
    const response = await api.get<ApiResponse<AmizadeResponse[]>>('/api/v1/amizades/enviadas');
    return response.data.dados;
  },

  async enviar(destinatarioId: number): Promise<AmizadeResponse> {
    const response = await api.post<ApiResponse<AmizadeResponse>>(
      `/api/v1/amizades/${destinatarioId}`
    );
    return response.data.dados;
  },

  async aceitar(amizadeId: number): Promise<AmizadeResponse> {
    const response = await api.put<ApiResponse<AmizadeResponse>>(
      `/api/v1/amizades/${amizadeId}/aceitar`
    );
    return response.data.dados;
  },

  async recusar(amizadeId: number): Promise<AmizadeResponse> {
    const response = await api.put<ApiResponse<AmizadeResponse>>(
      `/api/v1/amizades/${amizadeId}/recusar`
    );
    return response.data.dados;
  },

  async remover(amizadeId: number): Promise<void> {
    await api.delete(`/api/v1/amizades/${amizadeId}`);
  },

  async cancelar(amizadeId: number): Promise<void> {
    await api.delete(`/api/v1/amizades/${amizadeId}/cancelar`);
  },

  async buscar(termo: string): Promise<PerfilPublicoResponse[]> {
    const response = await api.get<ApiResponse<PerfilPublicoResponse[]>>(
      '/api/v1/participantes/buscar',
      { params: { termo } }
    );
    return response.data.dados;
  },

  async perfilPublico(id: number): Promise<PerfilPublicoResponse> {
    const response = await api.get<ApiResponse<PerfilPublicoResponse>>(
      `/api/v1/participantes/${id}/perfil`
    );
    return response.data.dados;
  },

  async perfilPublicoPorUsername(username: string): Promise<PerfilPublicoResponse> {
    const response = await api.get<ApiResponse<PerfilPublicoResponse>>(
      `/api/v1/participantes/by-username/${encodeURIComponent(username)}/perfil`
    );
    return response.data.dados;
  },

  async listarAmigosDeUsuario(userId: number): Promise<PerfilPublicoResponse[]> {
    const response = await api.get<ApiResponse<PerfilPublicoResponse[]>>(
      `/api/v1/participantes/${userId}/amigos`
    );
    return response.data.dados;
  },
};
